feat(posts): add blog post to expose the architecture of our web homologation stack (#480)
M6Web/m6web.github.io: _posts/2025-12-20-web-homologation-as-app.md

---
layout: post
title: "Building a Scalable, Multi-Tenant Homologation Stack at Bedrock Streaming"

> description: "We share here QA context,key takeaways and architecture decisions and reveal the architecture of the Web homologation stack."

Reviewer: ```suggestion
description: "We share here QA context, key takeaways and reveal the architecture of the Web homologation stack."
```

tags: [frontend, playwright, qa, web]
author: [m_bayard]
cover: /assets/images/
color: rgb(251,87,66)
language: en
feature-img: "/images/posts/"

> thumbnail: "/images/posts/"

Reviewer: Missing artifacts 🕳️

  - But also higher responsability regarding the quality of the injected tests and code overall 🥐. The fear of breaking things is natural but approved reviews and a green CI are always keeping up the quality level.

## 🏗️ Homologation Architecture Overview

Today this architecture allows writing one test that runs across 4 customers × 2 platforms × multiple environments.
The architecture follows a clean layered separation between business requirements (Gherkin), test logic (steps), and implementation details (utils).

1. Feature Files (BDD Layer)
   Path: features/@desktop/ & features/@webview/ Categorized by domain: @core (navigation), @player (VOD/Live), and @ulc (account/payment).

2. Step Definitions (Glue Layer)
   Path: features/steps/\*.steps.ts Standardized user actions (When) and assertions (Then) such as visibility.steps.ts or authentification.steps.ts.

3. Utilities (Implementation Layer)
   Path: features/steps/utils/\*.utils.ts The "brains" of the operation, including config.utils.ts for multi-tenant data retrieval and player.utils.ts for video interactions.

4. Configuration (Multi-Tenant Layer)

>    Path: config/{customer}/ Where the magic happens. This layer stores customer-specific account.config.ts, translations.config.ts, and environment URLs.

Reviewer: It is not displaying properly. Here is another try (but without any garantee):

```suggestion
**1. Feature Files (BDD Layer)**
Path: `features/@desktop` & `features/@webview`
Categorized by domain: `@core` (navigation), `@player` (VOD/Live), and `@ulc` (account/payment)

**2. Step Definitions (Glue Layer)**
Path: `features/steps/\*.steps.ts`
Standardized user actions (When) and assertions (Then) such as visibility.steps.ts or authentification.steps.ts.

**3. Utilities (Implementation Layer)**
Path: `features/steps/utils/\*.utils.ts`
The "brains" of the operation, including functions for multi-tenant data retrieval and video interactions

**4. Configuration (Multi-Tenant Layer)**
Path: `config/{customer}`
Where the magic happens. This layer stores customer-specific account.config.ts, translations.config.ts, and environment URLs.
```

```
apps/homologation/
├── features/                    # Gherkin feature files organized by platform
│   ├── @desktop/                # Desktop-specific test scenarios
│   ├── @webview/                # Mobile-specific test scenarios (same structure)
│   └── steps/                   # TypeScript step definitions
│       ├── authentification.steps.ts
│       ├── interactivity.steps.ts
│       ├── visibility.steps.ts
│       │── ...
│       └── utils/               # Utility functions for steps (getValue() for multi-tenant config)
│
├── config/                    # Customer-specific configuration
│   ├── config.ts              # Main config aggregator
│   ├── m6web/                 # M6+ specific config
│   ├── .../
│   ├── .../

> │   └── clientN/                  # client N specific config

Reviewer: Don't include real customer code

```suggestion
│   ├── .../
│   ├── .../
│   └── clientN/                  # client N specific config
```

- CI/CD In Action: Automated tests are integrated into our CI pipeline, ensuring quality gates are enforced before any release proceeds.
- The Human Factor: We’ve transformed the QA role. By transitioning to "Quality Architects," our engineers now own the automated patrimony, supported by targeted training to bridge the gap between manual testing and code.

## Next

All in all, this new stack is the backbone for delivering high-quality streaming experiences quickly.

The architecture and overall approach are confirmed to be successful, leading to a reduced and continuously improving homologation time as test coverage increases. We've implemented new rituals to thoroughly analyze results and maintain a strategic focus on both manual exploratory testing and executing more complex tests.

**Key takeaways for the QA community:**

- Embed tests in your codebase

- BDD approach is great to align teams on the “What” and is AI future-proof

- More automated testing means more time dedicated to valuable and strategic work from QA engineers (and less repetitive tasks)

**What's next:**

More AI and bring the automated testing to the age of agentic.

> Merge the testing approach with our developers.

Reviewer: ```suggestion
**What's next:**
- More AI and bring the automated testing to the age of agentic.
- Merge the testing approach with our developers.
```

---

## ⚙️ Building Blocks of the New Homologation Stack

Our solution hinges on a deliberate choice of tools and methodologies designed for readability, maintainability, and future-proofing:

**Core Stack**

| Component                                                      | Role                      | Key Benefit                                                                                                                         |
| -------------------------------------------------------------- | ------------------------- | ----------------------------------------------------------------------------------------------------------------------------------- |
| [Playwright](https://playwright.dev/)                          | Browser Automation Engine | Speed and Reliability: Unified API for all major browsers (Chromium, Firefox, WebKit), mobile emulation, and excellent DX.          |
| [playwright-bdd](https://vitalets.github.io/playwright-bdd/#/) | BDD Framework (Gherkin)   | Collaboration and Readability: Enables tests to be written in a natural, shared language while managing Playwright test generation. |

> Note: our dictionary of steps is greatly inspired by [uuv](https://github.com/Orange-OpenSource/uuv) which aims to provides an ecosystem that simplifies the writing of End-to-End tests in a BDD approach and a user-centric way and accessibility-first selectors.

Reviewer: To have a more dynamic layout like 
<img width="227" height="155" alt="Screenshot 2025-12-18 at 16 36 12" src="https://github.com/user-attachments/assets/d47a1b59-3929-459e-bdbb-11543fa82602" />

```suggestion
> ℹ️ Note: our dictionary of steps is greatly inspired by [uuv](https://github.com/Orange-OpenSource/uuv) which aims to provides an ecosystem that simplifies the writing of End-to-End tests in a BDD approach and a user-centric way and accessibility-first selectors.
```

**Requirements and rationale behind**

- Natural Language BDD (Gherkin): This is a key enabler for collaboration.

>   - Easy Review: Allows QA Engineers, Product Owners (POs), and Developers (DEVs) to easily read, review, and contribute to the test specifications without needing deep coding knowledge.

Reviewer: ```suggestion
  - **Easy Review**: Allows QA Engineers, Product Owners (POs), and Developers (DEVs) to easily read, review, and contribute to the test specifications without needing deep coding knowledge.
```

  - Leverage our CI environments and workflows for future integration with deployment.

  - But also higher responsability regarding the quality of the injected tests and code overall 🥐. The fear of breaking things is natural but approved reviews and a green CI are always keeping up the quality level.

## 🏗️ Homologation Architecture Overview

Today this architecture allows writing one test that runs across 4 customers × 2 platforms × multiple environments.
The architecture follows a clean layered separation between business requirements (Gherkin), test logic (steps), and implementation details (utils).

**1. Feature Files (BDD Layer)**
Path: `features/@desktop/` & `features/@webview/`
Categorized by domain: `@core` (navigation), `@player` (VOD/Live), and `@ulc` (account/payment).

**2. Step Definitions (Glue Layer)**
Path: `features/steps/\*.steps.ts`
Standardized user actions (When) and assertions (Then) such as visibility.steps.ts or authentification.steps.ts.

**3. Utilities (Implementation Layer)**
Path: `features/steps/utils/\*.utils.ts`

> The "brains" of the operation, including config.utils.ts for multi-tenant data retrieval and player.utils.ts for video interactions.

Reviewer: I don't think it's relevant to mention the filename `config.utils.ts` and `player.utils.ts` which can be quite confusing for the reader. I suggest:
```suggestion
The "brains" of the operation, including some functions for multi-tenant data retrieval and video interactions.
```

**4. Configuration (Multi-Tenant Layer)**
Path: `config/{customer}/`

> Where the magic happens. This layer stores customer-specific account.config.ts, translations.config.ts, and environment URLs.

Reviewer: Layout still not OK
<img width="586" height="299" alt="Screenshot 2025-12-18 at 16 37 45" src="https://github.com/user-attachments/assets/397702c6-533d-4125-baec-ff881840f267" />

Try this:
```suggestion
**1. Feature Files (BDD Layer)**

Path: `features/@desktop/` & `features/@webview/`

Categorized by domain: `@core` (navigation), `@player` (VOD/Live), and `@ulc` (account/payment).

**2. Step Definitions (Glue Layer)**

Path: `features/steps/\*.steps.ts`

Standardized user actions (When) and assertions (Then) such as visibility.steps.ts or authentification.steps.ts.

**3. Utilities (Implementation Layer)**

Path: `features/steps/utils/\*.utils.ts`

The "brains" of the operation, including config.utils.ts for multi-tenant data retrieval and player.utils.ts for video interactions.

**4. Configuration (Multi-Tenant Layer)**

Path: `config/{customer}/`

Where the magic happens. This layer stores customer-specific account.config.ts, translations.config.ts, and environment URLs.
```

│   ├── selectors.ts           # Main selectors aggregator
│   ├── clientA/               # client A specific config
│   ├── .../
│   ├── .../
│   └── clientN/                  # client N specific config
│       ├── datas.config.ts       # Test data
│       ├── urls.config.ts        # URL mappings
│       └── translations.config.ts # Merged translations from different source
```

### 🔄 Data Flow Example: Multi-Tenant Magic

When executing a test like:

```js
Then the user should see the text "translation.account.menu.logout"
```

Step receives the abstract key `"translation.account.menu.logout"`.

Step calls a utils fonction `getValue("translation.account.menu.logout")`.

Config resolves to the specific customer's value (e.g., `"Kijelentkezés"` for one customer, `"Logout"` for another).

Playwright locates the element with that specific text and validates visibility.

## ✦ The Future: AI Readiness and CI Integration

- AI and BDD: The structured, human-readable format of Gherkin makes it an ideal input for future AI agents.
- CI/CD In Action: Automated tests are integrated into our CI pipeline, ensuring quality gates are enforced before any release proceeds.

> - The Human Factor: We’ve transformed the QA role. By transitioning to "Quality Architects," our engineers now own the automated patrimony, supported by targeted training to bridge the gap between manual testing and code.

Reviewer: ```suggestion
- **AI and BDD**: The structured, human-readable format of Gherkin makes it an ideal input for future AI agents.
- **CI/CD In Action**: Automated tests are integrated into our CI pipeline, ensuring quality gates are enforced before any release proceeds.
- **The Human Factor**: We’ve transformed the QA role. By transitioning to "Quality Architects," our engineers now own the automated patrimony, supported by targeted training to bridge the gap between manual testing and code.
```
